ships/0026-shipwright-managed-push.md

```diff
@@ -0,0 +1,149 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: shipwright-managed-push
+authors:
+  - @SaschaSchwarze0
+reviewers:
+  - @HeavyWombat
+  - @GabeMontero
+  - @ImJasonH
+approvers:
+  - @adambkaplan
+creation-date: 2022-01-19
+last-updated: 2022-01-19
+status: implementable
+---
+
+# Make the push operation to the container registry a shipwright-managed operation
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+None
+
+## Summary
+
+This ship describes why and how we provide means for build strategies to store the produced container image as a tarball in ephemeral storage instead of doing the image push. The push will then be performed by a shipwright-managed step.
+
+## Motivation
+
+Shipwrights mission is to turn source code into container images and push them to a container registry. So far, every build strategy author is responsible to implement the push operation. In most cases this is straight-forward because all of the tools that we use in our sample build strategies support this.
+
+It became more complex when we introduced results to pass information about the image to the BuildRun status: the digest and the size. In many cases, additional script-style logic had to be added to build strategies to parse standardized or propriatary files to access this information, and then to write it to response files. This made those strategies more complex.
+
+Another feature introduced another pain: we allow to specify labels and annotations in the Build and BuildRun to be added to the image. With the image being pushed already by the strategy step, this feature was only implementing by re-pushing the augmented image - potentially causing issues in scenarios where the image push triggers other logic such as deployments.
```

Comment:
The ship outlines in total five addressed pain points or future scenarios:

- simplifying build strategies by removing the need to take care of results
- getting rid of the duplicate push
- consistent handling of insecure registries because of a single code path that performs the image push
- security scan before a push supported by shipwright
- creating a sbom of the produced image supported by shipwright

Image signing as mentioned is not necessarily needing this ship because signing happens after the push.

All these things should be weighted against the necessary changes for a fair comparison.

And I see this list extensible:

- retries done in our image push code
- making the image push eventually optional by making the the whole output section in a Build/BuildRun optional, useful in test scenarios where the image build is only tested to be passing and to produce a secure image, but the image is not actually needed

Reply:
Fair. I think if we have room to adopt best practices(e.g. sbom for supply chain) , security scans or improvements, I do not see why we shouldn't do this. Signing will be important, to complement this story.

On

> making the image push eventually optional by making the the whole output section in a Build/BuildRun optional, useful in test scenarios where the image build is only tested to be passing and to produce a secure image, but the image is not actually needed

this might contradict the conceptual integrity of a Build, definitely something not for here.